<!DOCTYPE html>
<!--[if lt IE 7 ]> <html class="no-js ie6"> <![endif]-->
<!--[if IE 7 ]>    <html class="no-js ie7"> <![endif]-->
<!--[if IE 8 ]>    <html class="no-js ie8"> <![endif]-->
<!--[if IE 9 ]>    <html class="no-js ie9"> <![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--> <html class="no-js"> <!--<![endif]-->
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1">
    
      
        <title>Getting Started - HyPhy - Hypothesis Testing using Phylogenies</title>
      
      
      
        <link rel="canonical" href="http://squidfunk.github.io/mkdocs-material/getting-started/">
      
      
        <meta name="author" content="VEG">
      
    
    <meta property="og:url" content="http://squidfunk.github.io/mkdocs-material/getting-started/">
    <meta property="og:title" content="HyPhy - Hypothesis Testing using Phylogenies">
    <meta property="og:image" content="http://squidfunk.github.io/mkdocs-material/getting-started//../images/logo.svg">
    <meta name="apple-mobile-web-app-title" content="HyPhy - Hypothesis Testing using Phylogenies">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    
      <link rel="apple-touch-icon" href="../images/logo.svg">
    
    
    <link rel="shortcut icon" type="image/x-icon" href="../assets/images/favicon.ico">
    <link rel="icon" type="image/x-icon" href="../assets/images/favicon.ico">
    <style>
      @font-face {
      	font-family: 'Icon';
      	src: url('../assets/fonts/icon.eot?52m981');
      	src: url('../assets/fonts/icon.eot?#iefix52m981')
               format('embedded-opentype'),
      		   url('../assets/fonts/icon.woff?52m981')
               format('woff'),
      		   url('../assets/fonts/icon.ttf?52m981')
               format('truetype'),
      		   url('../assets/fonts/icon.svg?52m981#icon')
               format('svg');
      	font-weight: normal;
      	font-style: normal;
      }
    </style>
    <link rel="stylesheet" href="../assets/stylesheets/application.css">
    <link rel="stylesheet" href="../assets/stylesheets/index.css">
		<link rel="stylesheet" href="../assets/stylesheets/styles.css">
    <link rel="stylesheet" href="../assets/stylesheets/mobile.css">
    
    
      
      
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:400,700|Ubuntu+Mono">
      <style>
        body, input {
          font-family: 'Ubuntu', Helvetica, Arial, sans-serif;
        }
        pre, code {
          font-family: 'Ubuntu Mono', 'Courier New', 'Courier', monospace;
        }
      </style>
    
    
    <script src="../assets/javascripts/modernizr.js"></script>
    
  </head>
  
  
  
  <body class=" ">
    
    <div class="backdrop">
      <div class="backdrop-paper"></div>
    </div>
    <input class="toggle" type="checkbox" id="toggle-drawer">
    <input class="toggle" type="checkbox" id="toggle-search">
    <label class="toggle-button overlay" for="toggle-drawer"></label>
    <header class="header">
      <nav aria-label="Header">
  <div class="bar default">
    <div class="button button-menu" role="button" aria-label="Menu">
      <label class="toggle-button icon icon-menu" for="toggle-drawer">
        <span></span>
      </label>
    </div>
    <div class="stretch">
      <div class="header-title">
        <a href="#"><img src="images/header-logo.svg" alt="HyPhy Logo" id="header-logo" /></a>
        <a href="#" id="header-title">HyPhy<br />
        <span class="header-title-light">Hypothesis Testing using Phylogenies</span></a>
      </div>
      
      <div class="mobile-header-title">
        <a href="#"><img src="images/header-logo.svg" alt="HyPhy Logo" id="header-logo" /></a>
        <a href="#">HyPhy</a>
      </div>
    </div>
    <div class="button button-search" role="button" aria-label="Search">
      <label class="toggle-button icon icon-search" title="Search" for="toggle-search"></label>
    </div>
  </div>
  <div class="bar search">
    <div class="button button-close" role="button" aria-label="Close">
      <label class="toggle-button icon icon-back" for="toggle-search"></label>
    </div>
    <div class="stretch">
      <div class="field">
        <input class="query" type="text" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false">
      </div>
    </div>
    <div class="button button-reset" role="button" aria-label="Search">
      <button class="toggle-button icon icon-close" id="reset-search"></button>
    </div>
  </div>
</nav>
    </header>
    <main class="main">
      
      <div class="drawer">
        <nav aria-label="Navigation">
  
  <a href="/index.html" class="project">
    <div class="banner">
      
        <div class="logo">
          <img src="../images/logo.svg">
        </div>
      
    </div>
  </a>
  <div class="scrollable">
    <div class="wrapper">
      <div class="toc">
        <ul>
          
            <!-- Render sections -->


  <li>
    <a class=""
      title="Home" href=".." >
      Home
    </a>

    <!-- Expand active pages -->
    
  </li>


          
            <!-- Render sections -->


  <li>
    <a class=""
      title="Download" href="../register/" >
      Download
    </a>

    <!-- Expand active pages -->
    
  </li>


          
            <!-- Render sections -->



          
            <!-- Render sections -->


  <li>
    <a class=""
      title="Installation" href="../installation/" >
      Installation
    </a>

    <!-- Expand active pages -->
    
  </li>


          
            <!-- Render sections -->


  <li>
    <a class="current"
      title="Getting Started" href="./" >
      Getting Started
    </a>

    <!-- Expand active pages -->
    

      <!--
        The top-level anchor must be skipped if the article contains a h1
        headline, since it would be redundant to the link to the current page
        that is located just above the anchor. Therefore we directly continue
        with the children of the anchor.
      -->
      

      <!-- Render anchors of active page -->
      
        <ul>
          

            <!-- Render anchor -->
            <li class="anchor">
              <a title="What is HyPhy?" href="#what-is-hyphy">
                What is HyPhy?
              </a>
            </li>
          

            <!-- Render anchor -->
            <li class="anchor">
              <a title="Choosing a method for selection inference" href="#choosing-a-method-for-selection-inference">
                Choosing a method for selection inference
              </a>
            </li>
          
        </ul>
      
    
  </li>


          
            <!-- Render sections -->


  <li>
    <a class=""
      title="Resources" href="../resources/" >
      Resources
    </a>

    <!-- Expand active pages -->
    
  </li>


          
            <!-- Render sections -->

  <li>
    <span class="section">Methods</span>
    <ul>

      <!-- Render pages of section -->
      
        <!-- Render sections -->


  <li>
    <a class=""
      title="General Information" href="../methods/general/" >
      General Information
    </a>

    <!-- Expand active pages -->
    
  </li>


      
        <!-- Render sections -->


  <li>
    <a class=""
      title="Selection" href="../methods/selection-methods/" >
      Selection
    </a>

    <!-- Expand active pages -->
    
  </li>


      
    </ul>
  </li>

<!-- Render page link -->

          
            <!-- Render sections -->

  <li>
    <span class="section">Tutorials</span>
    <ul>

      <!-- Render pages of section -->
      
        <!-- Render sections -->


  <li>
    <a class=""
      title="Using HyPhy to Detect Selection" href="../tutorials/current-release-tutorial/" >
      Using HyPhy to Detect Selection
    </a>

    <!-- Expand active pages -->
    
  </li>


      
        <!-- Render sections -->


  <li>
    <a class=""
      title="v2.3 Quick Start" href="../tutorials/dev-branch-tutorial/" >
      v2.3 Quick Start
    </a>

    <!-- Expand active pages -->
    
  </li>


      
    </ul>
  </li>

<!-- Render page link -->

          
        </ul>
      </div>
    </div>
  </div>
</nav>
      </div>
      <article class="article">
        <div class="wrapper">
          <h2 id="what-is-hyphy">What is HyPhy?<a class="headerlink" href="#what-is-hyphy" title="Permanent link">#</a></h2>
<p>HyPhy (<strong>Hy</strong>pothesis Testing using <strong>Phy</strong>logenies) is an open-source software package for the analysis of genetic sequences (in particular the inference of natural selection) using techniques in phylogenetics, molecular evolution, and machine learning. It features a rich scripting language for limitless customization of analyses. Additionally, HyPhy features support for parallel computing environments (via message passing interface). HyPhy has over 8000 registered users and has been cited in over 900 peer-reviewed publications (Google Scholar). 
<!-- UPDATE THIS, also check citation counts above.: Continued development of HyPhy is currently supported in part by an NIGMS R01 award 1R01GM093939. --></p>
<p>You can use HyPhy in two ways:</p>
<ul>
<li>Run HyPhy our accompanying webserver <a href="http://datamonkey.org">Datamonkey</a> (or see the <a href="http://test.datamonkey.org">development version of Datamonkey</a> for some newer methods).</li>
<li>Run HyPhy from the command line on your local computer/server. Follow <a href="../installation/">these instructions</a> for download and installation. </li>
</ul>
<h2 id="choosing-a-method-for-selection-inference">Choosing a method for selection inference<a class="headerlink" href="#choosing-a-method-for-selection-inference" title="Permanent link">#</a></h2>
<p>HyPhy provides a suite of diverse phylogenetic methodologies for testing specific hypotheses about selection in protein-coding and/or amino-acid multiple sequence alignments. Which method you select will depend on your specific question. Below we recommend several methods for different purposes, linked to more in depth descriptions. Tutorials for using these methods are also available <a href="../tutorials/current-release-tutorial">here</a>.</p>
<h3 id="are-individual-sites-subject-to-pervasive-across-the-whole-phylogeny-positive-or-purifying-selection">Are individual sites subject to <em>pervasive</em> (across the whole phylogeny) positive or purifying selection?<a class="headerlink" href="#are-individual-sites-subject-to-pervasive-across-the-whole-phylogeny-positive-or-purifying-selection" title="Permanent link">#</a></h3>
<ul>
<li><a href="../methods/selection-methods/#fel">FEL</a> (<strong>F</strong>ixed <strong>E</strong>fects <strong>L</strong>ikelihood) is suitable for small-to-medium sized data sets.</li>
<li><a href="../methods/selection-methods/#slac">SLAC</a> (<strong>S</strong>ingle-<strong>L</strong>ikelihood <strong>A</strong>ncestor <strong>C</strong>ounting) is an approximate method with accuracy similar to FEL, but suitable for larger datasets. However, SLAC is not suitable for highly-diverged sequences.</li>
<li><a href="../methods/selection-methods/#fubar">FUBAR</a> (<strong>F</strong>ast, <strong>U</strong>nconstrained <strong>B</strong>ayesian <strong>A</strong>pp<strong>R</strong>oximation) is suitable for medium-to-large data sets and is expected to have more power than FEL for detecting pervasive selection at sites. <strong>FUBAR is the preferred approach for inferring pervasive selection.</strong></li>
</ul>
<h3 id="are-individual-sites-subject-to-episodic-at-a-subset-of-branches-positive-or-purifying-selection">Are individual sites subject to <em>episodic</em> (at a subset of branches) positive or purifying selection?<a class="headerlink" href="#are-individual-sites-subject-to-episodic-at-a-subset-of-branches-positive-or-purifying-selection" title="Permanent link">#</a></h3>
<ul>
<li><a href="../methods/selection-methods/#meme">MEME</a> (<strong>M</strong>ixed <strong>E</strong>ffects <strong>M</strong>odel of <strong>E</strong>volution) tests for episodic selection at individual sites. Note that MEME does not accept <em>a priori</em> branch specifications. <strong>MEME is the preferred approach for detecting positive selection at individual sites.</strong></li>
<li><a href="../methods/selection-methods/#absrel">aBSREL</a> (<strong>a</strong>daptive <strong>B</strong>ranch-<strong>S</strong>ite <strong>R</strong>andom <strong>E</strong>ffects <strong>L</strong>ikelihood) is an improved version of the common "branch-site" class of models. aBSREL allows either for <em>a priori</em> specification of branch(es) to test for selection, or can test each lineage for selection in an exploratory fashion. Note that the exploratory approach will sacrifice power.</li>
</ul>
<h3 id="has-a-gene-experienced-positive-selection-at-any-site-on-a-particular-branch-or-set-of-branches">Has a gene experienced positive selection at any site on a particular branch or set of branches?<a class="headerlink" href="#has-a-gene-experienced-positive-selection-at-any-site-on-a-particular-branch-or-set-of-branches" title="Permanent link">#</a></h3>
<ul>
<li><a href="../methods/selection-methods/#busted">BUSTED</a> (<strong>B</strong>ranch-<strong>S</strong>ite <strong>U</strong>nrestricted <strong>S</strong>tatistical <strong>T</strong>est for <strong>E</strong>pisodic <strong>D</strong>iversification) will test for gene-wide selection at pre-specified lineages. This method is particularly useful for relatively small datasets (fewer than 10 taxa) where other methods may not have sufficient power to detect selection. <strong>This method is not suitable for identifying specific sites subject to positive seleciton.</strong></li>
</ul>
<h3 id="has-gene-wide-selection-pressure-been-relaxed-or-intensified-along-a-certain-subset-of-branches">Has gene-wide selection pressure been relaxed or intensified along a certain subset of branches?<a class="headerlink" href="#has-gene-wide-selection-pressure-been-relaxed-or-intensified-along-a-certain-subset-of-branches" title="Permanent link">#</a></h3>
<ul>
<li><a href="../methods/selection-methods/#relax">RELAX</a> tests for a relaxation (e.g. where purifying selection has become less stringent) or an intensification (e.g. where purifying selection has become stronger) of selection pressures along a specified set of "test" branches. <strong>This method is not suitable for detecting positive selection.</strong></li>
</ul>
<!--
### Are individual sites within a gene subject to *directional* selection, i.e. selection pressure to evolve towards a specific set of amino acids?
* [FADE](./methods/selection-methods/#fade) tests for directional selection at specific sites in *protein* alignments.
-->
			 <!--
          <aside class="copyright" role="note">
            
            Documentation built with
            <a href="http://www.mkdocs.org" target="_blank">MkDocs</a>
            using the
            <a href="http://squidfunk.github.io/mkdocs-material/" target="_blank">
              Michael Li's Custom Theme
            </a>
            theme.
          </aside>
				-->
          
            <footer class="footer">
              
  <nav class="pagination" aria-label="Footer">
    <div class="previous">
      
        <a href="../installation/" title="Installation">
          <span class="direction">
            Previous
          </span>
          <div class="page">
            <div class="button button-previous" role="button" aria-label="Previous">
              <i class="icon icon-back"></i>
            </div>
            <div class="stretch">
              <div class="title">
                Installation
              </div>
            </div>
          </div>
        </a>
      
    </div>
    <div class="next">
      
        <a href="../resources/" title="Resources">
          <span class="direction">
            Next
          </span>
          <div class="page">
            <div class="stretch">
              <div class="title">
                Resources
              </div>
            </div>
            <div class="button button-next" role="button" aria-label="Next">
              <i class="icon icon-forward"></i>
            </div>
          </div>
        </a>
      
    </div>
  </nav>


<div class="container citation-footer">
  <p>HyPhy is funded jointly by <a
                 href="https://projectreporter.nih.gov/project_info_details.cfm?aid=9102131"
                 target="_blank">MIDAS</a> and <a
                 href="https://projectreporter.nih.gov/project_info_description.cfm?aid=8886417"
                 target="_blank">NIH award R01 GM093939</a></p>
</div>



            </footer>
          
        </div>
      </article>
      <div class="results" role="status" aria-live="polite">
        <div class="scrollable">
          <div class="wrapper">
            <div class="meta"></div>
            <div class="list"></div>
          </div>
        </div>
      </div>
    </main>
    <script>
      var base_url = '..';
      var repo_id  = '';
    </script>
    <script src="../assets/javascripts/application.js"></script>
    
      <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    
      <script src="../assets/mathjaxhelper.js"></script>
    
    
				<!-- Piwik -->
				<script type="text/javascript">
					var _paq = _paq || [];
					_paq.push(["setDomains", ["*.hyphy.org"]]);
					_paq.push(['trackPageView']);
					_paq.push(['enableLinkTracking']);
					(function() {
						var u="//datamonkey.org/piwik/";
						_paq.push(['setTrackerUrl', u+'piwik.php']);
						_paq.push(['setSiteId', 10]);
						var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
						g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
					})();
				</script>
				<noscript><p><img src="//datamonkey.org/piwik/piwik.php?idsite=10" style="border:0;" alt="" /></p></noscript>
				<!-- End Piwik Code -->
    
  </body>
</html>